src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5377,20 +5377,112 @@ private int GetNullableOfTValueSlot(TypeSymbol containingType, int containingSlo
 
         protected override void VisitForEachExpression(BoundForEachStatement node)
         {
-            var expr = node.Expression;
-            // Need to remove any implicit conversions from this expression and call ApplyConversion
-            // ourselves. https://github.com/dotnet/roslyn/issues/35151
-            VisitRvalue(expr);
+            if (node.Expression.Kind != BoundKind.Conversion)
+            {
+                // If we're in this scenario, there was a binding error, and we should suppress any further warnings.
+                Debug.Assert(node.HasErrors);
+                VisitRvalue(node.Expression);
+                return;
+            }
+
+            var (expr, conversion) = RemoveConversion(node.Expression, includeExplicitConversions: false);
+
+            // There are 7 ways that a foreach can be created:
```

Comment:
>There are 7 ways that a foreach can be created: [](start = 15, length = 47)

This may be overkill, but I'm wondering if we should just have an enum value that we save during initial binding.
We do something like this when analyzing `??` or maybe ` ? : ` (where initial binding knows whether we used the type from the left or right, or some other dynamic case).

Reply:
After consideration, I think I won't do this, and stick with what we have for now.

---
In reply to: [279560894](https://github.com/dotnet/roslyn/pull/35316#discussion_r279560894) [](ancestors = 279560894)